### PR TITLE
[TorchFX] Graph/metatypes tests implementation

### PR DESCRIPTION
**Pull request by anzr299**
### Changes

1. Added pytest to check if all the ops in the generated graph for resnet18, mobilenet_v3_small, vit_b_16, swin_v2_s and yolov8n models have valid metatypes in NNCF
2. Added pytest to check if the generated NNCF graph for resnet18, mobilenet_v3_small, vit_b_16, swin_v2_s and yolov8n match with reference graphs. (tests/torch/fx/test_models.py)
3. Add metatype mapping for the following:
 - linalg_vector_norm
 - clamp
 - clamp_min
 - hardswish_
 - dropout_

### Reason for changes

Check if all model ops are supported by NNCF metatypes

### Related tickets

#2778

### Tests

New test file test_models.py added in tests/torch/fx/test_models.py

### CODE REVIEW

File: tests/torch/fx/test_models.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            return data_item[0].to(device)
+
+        calibration_dataset = nncf.Dataset(calibration_dataset, transform_fn)
+
+        with torch.no_grad():
+            ex_input = next(iter(calibration_dataset.get_inference_data()))
+            model.eval()
+            exported_model = capture_pre_autograd_graph(model, args=(ex_input,))
+            nodes = retrieve_nodes(exported_model)
+            for node in nodes:
```
Comment (alexsu52) [inline tests/torch/fx/test_models.py:100]: I suggest to compare with reference graph. The example is https://github.com/openvinotoolkit/nncf/blob/f02e5efe947fb968d6d7bee44ab5d392a4e92564/tests/openvino/native/test_nncf_graph_builder.py#L51

File: tests/torch/fx/test_models.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        return False
+    return True
+
+def retrieve_nodes(model: torch.fx.GraphModule):
+    for node in model.graph.nodes:
+        yield node
+
+
+@pytest.mark.parametrize("test_case", MODELS)
+def test_sanity(test_case: ModelCase, tiny_imagenet_dataset):
```
Comment (alexsu52) [inline tests/torch/fx/test_models.py]: @daniil-lyakhov Could you clarify what kind test do you request to add for metatypes? Could you provide a link to example of such test?

File: tests/torch/fx/test_models.py (hunk 1 above)
Comment (anzr299) [reply tests/torch/fx/test_models.py:100]: Alright, I will have a look into it.

File: tests/torch/fx/test_models.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        node_type = str(node.target).split(".")[1]
+        return node_type
+    return ""
+
+
+def isNodeMetatype(node_type: str) -> bool:
+    op_type = PT_OPERATOR_METATYPES.get_operator_metatype_by_op_name(node_type)
+    if op_type is UnknownMetatype:
+        return False
+    return True
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:80]: Please use underscore in  function names as it mentioned in the [styleguide](https://github.com/openvinotoolkit/nncf/blob/develop/docs/styleguide/PyGuide.md#472-naming-conventions)  + minor refactoring
```suggestion
def is_node_metatype(node_type: str) -> bool:
    op_type = PT_OPERATOR_METATYPES.get_operator_metatype_by_op_name(node_type)
    return not op_type is UnknownMetatype:
```

File: tests/torch/fx/test_models.py (hunk 2 above)
Comment (daniil-lyakhov) [reply tests/torch/fx/test_models.py]: Similar test could be found [here](https://github.com/openvinotoolkit/nncf/blob/f02e5efe947fb968d6d7bee44ab5d392a4e92564/tests/openvino/native/test_nncf_graph_builder.py#L47-L52) (test_compare_nncf_graph_real_models) reference graphs should be added to the tests/torch/data directory

File: tests/torch/fx/test_models.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+
+def get_ref_metatypes_from_json(
+    model_name: str, model_metatypes: Dict[NNCFNodeName, Type[OperatorMetatype]]
+) -> Dict[NNCFNodeName, Type[OperatorMetatype]]:
+    model_json_name = get_json_filename(model_name)
+    complete_path = get_full_path_to_json(model_json_name)
+    check_to_create_ref(model_metatypes, complete_path)
+    model_ref_metatypes = read_json_file(complete_path)
+    return model_ref_metatypes
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:114]: Please decrease amount of functions, like this:
```suggestion


def get_ref_metatypes_from_json(
    model_name: str, model_metatypes: Dict[NNCFNodeName, Type[OperatorMetatype]]
) -> Dict[NNCFNodeName, Type[OperatorMetatype]]:
    model_json_name = get_json_filename(model_name)
    complete_path = get_full_path_to_json(model_json_name)
    
    json_parent_dir = Path(json_dir).parent

    if os.getenv("NNCF_TEST_REGEN_JSON") is not None:
        if not os.path.exists(json_parent_dir):
            os.makedirs(json_parent_dir)
        with safe_open(path, "w") as file:
        json.dump(metatype_dict, file)
    
     with safe_open(path, "r") as file:
        return json.load(file)
```

File: tests/torch/fx/test_models.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+
+MODELS = (
+    ModelCase("resnet18", (1, 3, 224, 224)),
+    ModelCase("mobilenet_v3_small", (1, 3, 224, 224)),
+    ModelCase("vit_b_16", (1, 3, 224, 224)),
+    ModelCase("swin_v2_s", (1, 3, 224, 224)),
+    ModelCase("yolov8", (1, 3, 224, 224)),
+)
+
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:60]: Do model builder instead
```
def torchvision_model_builder(model_id: str, ...):
    ...
       
def ultralytics_model_builder(model_id: str, ...):
       ...
       # run fist inference of the model here
       
TEST_MODELS = (torchvision_model_builder("resnet18", ...), ..., ultralytics_model_builder("yolov8n.yaml", ...))
```

File: tests/torch/fx/test_models.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        torch.manual_seed(42)
+        device = torch.device("cpu")
+        model_name = test_case.model_id
+        model = get_model(model_name, device)
+
+        with torch.no_grad():
+            ex_input = torch.randn(test_case.input_shape)
+            path_to_dot = fx_dir
+            model.eval()
+            model(ex_input)  # inferring from model to avoid anchor mutation in YOLOv8
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:143]: Please move it to the ultralytics model builder

File: tests/torch/fx/test_models.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    check_graph(model, dot_filename, path_to_dot)
+
+
+def compare_nncf_metatypes_model(model_metatypes: Dict[NNCFNodeName, Type[OperatorMetatype]], model_name: str):
+    ref_metatypes = get_ref_metatypes_from_json(model_name, model_metatypes)
+    if ref_metatypes == model_metatypes:
+        model_metatypes_match = True
+    else:
+        model_metatypes_match = False
+    return model_metatypes_match
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:128]:
```suggestion
```

File: tests/torch/fx/test_models.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            ex_input = torch.randn(test_case.input_shape)
+            path_to_dot = fx_dir
+            model.eval()
+            model(ex_input)  # inferring from model to avoid anchor mutation in YOLOv8
+            exported_model = capture_pre_autograd_graph(model, args=(ex_input,))
+            nncf_graph = GraphConverter.create_nncf_graph(exported_model)
+            compare_nncf_graph_model(nncf_graph, model_name, path_to_dot)
+            model_metatypes = {n.node_name: n.metatype.name for n in nncf_graph.get_all_nodes()}
+            model_metatypes_match = compare_nncf_metatypes_model(model_metatypes, model_name)
+            assert model_metatypes_match
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:149]:
```suggestion
            ref_metatypes = get_ref_metatypes_from_json(model_name, model_metatypes)
            assert model_metatypes == ref_metatypes
```

File: tests/torch/requirements.txt [hunk 9]
```diff
@@ -19,3 +19,9 @@ datasets==2.14.7
 evaluate==0.3.0
 openvino
 timm==0.9.2
+
+
+# Required for torch/fx tests
+torchvision
+fastdownload==0.0.7
+ultralytics
```
Comment (daniil-lyakhov) [inline tests/torch/requirements.txt:27]: Please specify version of the ultralytics and add comment: 
```
# TODO(dlyakhov) move ultralytics requirements to the nightly test
```

File: tests/torch/fx/test_models.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+@pytest.mark.parametrize("test_case", MODELS)
+def test_models(test_case: ModelCase, fx_dir):
+    with disable_patching():
+        torch.manual_seed(42)
+        device = torch.device("cpu")
+        model_name = test_case.model_id
+        model = get_model(model_name, device)
+
+        with torch.no_grad():
+            ex_input = torch.randn(test_case.input_shape)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:140]:
```suggestion
            ex_input = torch.ones(test_case.input_shape)
```

File: tests/torch/fx/test_models.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        model_metatypes_match = True
+    else:
+        model_metatypes_match = False
+    return model_metatypes_match
+
+
+@pytest.mark.parametrize("test_case", MODELS)
+def test_models(test_case: ModelCase, fx_dir):
+    with disable_patching():
+        torch.manual_seed(42)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:134]:
```suggestion
```

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]

File: tests/torch/fx/test_models.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+def get_full_path_to_json(model_json_name: str) -> str:
+    path_to_dir = TEST_ROOT / "torch" / "data" / "reference_graphs" / "fx" / "reference_metatypes"
+    path_to_json = path_to_dir / model_json_name
+    return path_to_json
+
+
+def read_json_file(path: Path):
+    with safe_open(path, "r") as file:
+        model_ref_metatypes = json.load(file)
+    return model_ref_metatypes
```
[the remark this answers is not in the document]
Comment (anzr299) [reply tests/torch/fx/test_models.py:95]: Done!

**Review by alexsu52 — APPROVED** [review]
LGTM.

**Comment by daniil-lyakhov** [comment]
@anzr299, please rebase your changes

**Comment by anzr299** [comment]
> @anzr299, please rebase your changes

Done

File: tests/torch/requirements.txt [hunk 13]
```diff
@@ -24,3 +24,4 @@ timm==0.9.2
 # Required for torch/fx tests
 torchvision
 fastdownload==0.0.7
+ultralytics==8.2.56  # TODO(dlyakhov) move ultralytics requirements to the nightly test
```
Comment (alexsu52) [inline tests/torch/requirements.txt:27]: Could you think about how to avoid adding `ultralytics` as dependency for pre-commit tests.

Comment (anzr299) [reply tests/torch/requirements.txt]: Done in latest merge from @daniil-lyakhov.

File: nncf/torch/graph/operator_metatypes.py [hunk 14]
```diff
@@ -1101,6 +1101,24 @@ class PTScaledDotProductAttentionMetatype(PTOperatorMetatype):
...
+    name = "VectorNormOp"
+    module_to_function_names = {NamespaceTarget.ATEN: ["linalg_vector_norm"]}
+
+
+@PT_OPERATOR_METATYPES.register()
+class PTClampMetatype(PTOperatorMetatype):
+    name = "ClampOp"
+    module_to_function_names = {NamespaceTarget.TORCH: ["clamp", "clamp_min"]}
+
+
```
Comment (daniil-lyakhov) [inline nncf/torch/graph/operator_metatypes.py:1121]: As this metatypes don't have any specific affect on any algorithm, I suggest to remove them. In that case following operation will have `UnknownMetatype` and will be treated the same way as now.

Comment (anzr299) [reply nncf/torch/graph/operator_metatypes.py:1121]: That sounds good, Alright.

Comment (anzr299) [reply nncf/torch/graph/operator_metatypes.py:1121]: Done!

Comment (daniil-lyakhov) [reply nncf/torch/graph/operator_metatypes.py]: Don't forget to update references as well

Comment (anzr299) [reply nncf/torch/graph/operator_metatypes.py]: Done!

**Review by daniil-lyakhov — APPROVED** [review]

**Pull request merged**